### ISSUE TITLE
Pattern advice for render/main plugin loading

### ISSUE THREAD

**Issue by Allow2CEO** [pksunkara/electron-plugin-manager#82]
Hi Pavan. So having a good time working out my use case. I am trying to build a pluggable framework for an electron app. The intention is to allow people to publish extensions that will have both a main process and a render process component. I have prototyped the load using your module and it's working well.

Now I need to work out how to best facilitate loading/unloading in sync across both processes. So the render side is a react view/display component, which is passed props containing redux sub-state (managed by the main app, which provides callbacks to the main app to update state). The render process is the UI for the plugin and allows users to interact with and configure that specific plugin.

The main app component provides the core functionality based on the state to interact with web services, network, printers, devices, etc. This is the engine side. The engine in the main process plugin part uses triggers off the main app and configurations for the specific plugin to provide the actual plugin functionality.

So the plugin in a single library that I pull in to both the main and the render processes to provide the engine and the interface. The main app acts as the conduit, so changes made to redux store by the ui help to control the engine side and vice-versa. The user can also disable/enable the plugin to "pause" it without uninstalling or deconfiguring it.

I trust the required steps are for me to manage the alignment of loaded/unloaded states between the engine side and the ui side, maybe with a configuration flag which drive appearance on the render side and the "liveness" of the engine side. I don't think it's worth unloading/loading the libraries at this stage in parallel with the enabled/disabled state, but maybe it's something I should consider?

Have you seen this use case (or considered it)? and If so, what advice/pointers could you share?

Also, I am having to not use es6 and jsx, but maybe I should use ondemand transpiling on render side component loading, but I'd need to know how to use your library to do this. I presume we need to extend or modify your library to achieve loading jsx and such at runtime? Have you thought about this? or has there been any demand?

Happy to work with you on any of this if it interests you.

Thanks,
Andrew

**Comment by pksunkara** [comment]
Regarding es6 and jsx, all you need to do is compile it to JS and make a package out of it and put it in plugins folder and it will behave like normal.

Regarding sync, as you can see from here, https://github.com/pksunkara/electron-plugin-manager/blob/master/lib/load.js#L5, it's just calling `require`, so both the functions are synchronous. You can use `ipc` to manage this loading on both at the same time. I would recommend load on `main.js` first. But this is something that is upto you design according to your app's needs.

**Comment by Allow2CEO** [comment]
Thanks, all makes sense and matches the approach I am taking.

**Comment by Allow2CEO** [comment]
Just on this, rather than opening a new ticket, can you comment on this?

I created my importable library:

```javascript
const TabContent = require('./Components/TabContent.js');

module.exports = function(context) {

    var wemo = {
        TabContent: TabContent
    };

    return wemo;
};
```

and TabContent is a react component:

```javascript
var React = require('react');
var createReactClass = require('create-react-class');

exports = module.exports = createReactClass({
   render: function() {
        return (
            React.createElement(div, {}, "yay")
        );
    }
});
```

and then I import it in the render process prior to displaying:

```javascript
const epm = require('electron-plugin-manager');
const dir = path.join(remote.app.getPath('appData'), 'allow2automate');

export default class Login extends Component {
    constructor(...args) {
        super(...args);
        const plugin = epm.load(dir, 'pluginName", remote.require);
        this.plugin = plugin({
            updateData: function(data) {
            }
        });
    }

   render() {
        const TabContent = this.plugin.TabContent;

        return (
            <TabContent {...plugin.data} />
        );
```

but it fails to render with an error:

Uncaught Error: Could not call remote function ''. Check that the function signature is correct. Underlying error: Cannot read property 'length' of undefined

if you examine this.plugin.TabContent, it's actually a proxy object, not the actual class.

Any tips/advice?
